android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/MenuPositioningData.kt

```diff
@@ -0,0 +1,269 @@
+
+    val appView = anchor.rootView
+    appView.getWindowVisibleDisplayFrame(displayFrame)
+
+    anchor.getLocationOnScreen(anchorPosition)
+
+    val distanceToLeft = anchorPosition[0] - displayFrame.left
+    val distanceToRight = displayFrame.right - (anchorPosition[0] + anchor.width)
+
+    return distanceToLeft to distanceToRight
+}
+
+/**
+ * Determine whether the container view can display all menu items (without scrolling) within
+ * the available height.
+ *
+ * @return The original container height if the container view can display all menu items
+ * (without scrolling), else calculate the maximum available container height for a scrollable
+ * view with a half menu item.
+ */
+private fun calculateContainerHeight(
+    recyclerViewHeight: Int,
+    recyclerViewItemCount: Int,
+    containerViewHeight: Int,
+    menuStylePadding: Int,
+    anchor: View,
+): Int {
+    // Get the total screen display height.
+    val totalHeight = anchor.rootView.measuredHeight
+
+    // Note: We cannot use getWindowVisibleDisplayFrame() as the height is dynamic based on whether
+    // the keyboard is open.
+    // Get any displayed system bars e.g. top status bar, bottom navigation bar or soft buttons bar.
+    val systemBars =
+        ViewCompat.getRootWindowInsets(anchor)?.getInsets(WindowInsetsCompat.Type.systemBars())
+    // Store the vertical status bars.
+    val topSystemBarHeight = systemBars?.top ?: 0
+    val bottomSystemBarHeight = systemBars?.bottom ?: 0
+
+    // Deduct any status bar heights from the total height.
+    val availableHeight = totalHeight - (bottomSystemBarHeight + topSystemBarHeight)
+
+    val menuItemHeight = recyclerViewHeight / recyclerViewItemCount
+
+    // We must take the menu container padding into account as this will be applied to the final height.
+    val containerPadding = containerViewHeight - recyclerViewHeight
+
+    val maxAvailableHeightForRecyclerView = availableHeight - containerPadding - menuStylePadding
+
+    // The number of menu items that can fit exactly (no cropping) within the max app height.
+    // Round the number of items to the closet Int value to ensure the max space available is utilized.
+    // E.g if 6.9 items fit, round to 7 so the calculation below will show 6.5 items instead of 5.5 .
+    val numberOfItemsFitExactly =
+        (maxAvailableHeightForRecyclerView.toFloat() / menuItemHeight.toFloat()).roundToInt()
+
+    val itemsAlreadyFitContainerHeight = recyclerViewItemCount <= numberOfItemsFitExactly
+
+    return if (itemsAlreadyFitContainerHeight) {
+        containerViewHeight
+    } else {
+        getCroppedMenuContainerHeight(numberOfItemsFitExactly, menuItemHeight, containerPadding)
+    }
+}
+
+private fun getAnimation(
+    fitsUp: Boolean,
+    fitsDown: Boolean,
+    drawingLeft: Boolean,
+    orientation: Orientation?,
+): Int {
+    val isUpOrientation = orientation == Orientation.UP
+    return when {
+        isUpOrientation && fitsUp -> if (drawingLeft) {
```

Comment:
This covers the edge case with the opened keyboard. UX team requested not to cut the menu in this case, but to show the menu from at the bottom of the screen.

Reply:
Thanks for the context here, may be worth including this as a comment on the function too 🙂